website/docs/language/providers/configuration.mdx

```diff
@@ -160,24 +235,138 @@ resource "aws_instance" "foo" {
 }
 ```
 
-Modules have some special requirements when passing in providers; see
-[The Module `providers` Meta-Argument](../../language/meta-arguments/module-providers.mdx)
-for more details. In most cases, only _root modules_ should define provider
-configurations, with all child modules obtaining their provider configurations
-from their parents.
+The `resource "aws_vpc" "private"` block uses `for_each` to declare one
+instance of the resource for each non-null element of `var.aws_regions`.
+The `provider` argument then uses `each.key` to select a different
+instance of `aws.by_region` for each instance of the resource, so that
+each would be declared in a different region.
+
+You can also choose a dynamic instance from a provider configuration for
+all resources in a child module as part of a `module` block. Refer to
+[Module instances with differing provider instances](../../language/meta-arguments/module-providers.mdx#module-instances-with-differing-provider-instances)
+for more information.
+
+Although the instance key expression in square brackets is dynamic, the
+provider configuration reference is static so that OpenTofu can infer the
+dependencies between `resource` blocks and `provider` blocks before
+evaluating any expressions. The dependencies then ensure that OpenTofu
+can resolve dynamic expressions in the correct order. This means that
+all instances of a particular resource must be bound to instances of the
+same provider configuration block, but can they each be bound to a different
+instance.
+
+:::warning
+**The `for_each` expression for a resource must not exactly match the
+`for_each` expression for its associated provider configuration.**
+
+OpenTofu uses a provider instance to plan and apply _all_ actions related
+to a resource instance, including destroying a resource instance that
+has been removed from the configuration.
+
+Therefore the provider instance associated with any resource instance must
+always remain in the configuration for at least one more plan/apply round
+after the resource instance has been removed, or OpenTofu will fail to
+plan to destroy the resource instance.
+
+The above example uses a null element in `var.aws_regions` to represent
+that a provider instance is needed but no resource instances should be
+associated with it.
+
+Setting a particular region's element to `null` would therefore cause
+OpenTofu to propose to destroy the `aws_vpc.private` instance for that
+region while retaining the provider instance needed to plan and apply that
+action. You can then remove the element altogether on the next round, once
+all of the associated resource instances have been destroyed.
+:::
+
+### Passing provider configurations between modules
+
+Each module has its own separate namespace of provider configurations, but
+it's possible for a parent module to pass some or all of its provider
+configurations into provider configuration addresses declared in a
+child module.
```

Comment:
Perhaps a note that we don't support provider instances passed into child modules without selection.

Reply:
This section is only here to connect with the docs about the `providers` argument in `module` blocks (linked below), which I think is the more natural home for that information.

As the moment that's only lightly _implied_ by the words on the page, rather than explicitly stated, but I'll add an extra paragraph to the other page to make that explicit.